space2stats_api/src/space2stats/h3_utils.py

```diff
@@ -1,67 +1,122 @@
-from itertools import chain
-from typing import Any, Dict, List, Optional
+from typing import Any, Dict, Iterable, List, Optional, Set
 
 import h3
 from shapely.geometry import MultiPolygon, Point, Polygon, mapping, shape
 
+# https://h3geo.org/docs/core-library/restable
+MAX_RESOLUTION = 15
+
+
+def _get_parent(
+    h3_ids: Iterable[str], resolution: int, target_resolution: int
+) -> Set[str]:
+    if resolution == target_resolution:
+        return set(h3_ids)
+
+    assert resolution > target_resolution
+    return {h3.h3_to_parent(h3_id, target_resolution) for h3_id in h3_ids}
+
+
+def _recursive_polyfill(
+    aoi_geojson: Dict[str, Any], resolution: int, original_resolution: int
+) -> Set[str]:
+    # Attempt to get H3 IDs at the current resolution
+    h3_ids = h3.polyfill(aoi_geojson, resolution, geo_json_conformant=True)
+
+    # If valid H3 IDs are found, return them
+    if h3_ids:
+        return _get_parent(h3_ids, resolution, original_resolution)
+
+    # If we haven't reached the maximum resolution, try the next higher resolution
+    if resolution < MAX_RESOLUTION:
+        return _recursive_polyfill(aoi_geojson, resolution + 1, original_resolution)
+
+    return set()
+
+
+def _find_within(aoi_geojson: Dict[str, Any], h3_ids: Iterable[str]) -> Set[str]:
+    # Convert AOI GeoJSON to Shapely geometry
+    aoi_shape = shape(aoi_geojson)
+
+    # Find H3 IDs that are within the polygon
+    contained_h3_ids = [
+        h3_id
+        for h3_id in h3_ids
+        if aoi_shape.contains(Polygon(h3.h3_to_geo_boundary(h3_id, geo_json=True)))
+    ]
+
+    return set(contained_h3_ids)
+
+
+def _find_touches(aoi_geojson: Dict[str, Any], h3_ids: Iterable[str]) -> Set[str]:
+    aoi_shape = shape(aoi_geojson)
+
+    neighbors = set()
+    for h3_id in h3_ids:
+        parent_id = h3.h3_to_parent(h3_id, h3.h3_get_resolution(h3_id) - 1)
+
+        for n_h3_id in h3.h3_to_children(
+            parent_id, h3.h3_get_resolution(parent_id) + 1
+        ):
+            boundary = Polygon(h3.h3_to_geo_boundary(n_h3_id, geo_json=True))
+            if h3.h3_indexes_are_neighbors(h3_id, n_h3_id) and aoi_shape.intersects(
+                boundary
+            ):
+                neighbors.add(n_h3_id)
+
+    return neighbors.union(h3_ids)
+
+
+def _generate_h3_ids(
+    aoi_geojson: Dict[str, Any], resolution: int, spatial_join_method: str
```

Comment:
we can define `spatial_join_method` with type `Literal["touche", "within", "centroid"]`

Reply:
great suggestion!